### ISSUE TITLE
Datepicker timezone issues

### ISSUE THREAD

**Issue by bradcavanagh** [angular-ui/angular-ui-OLDREPO#161]
I have distilled my issues into a fairly simple [jsFiddle](http://jsfiddle.net/AY82f/1/). Basically, I'm setting default values in my controller and binding them in the input tags, as well as in a simple display. The dates in the input fields come out as if the default dates are in UTC but displayed in my local timezone (since I'm in PDT, they're coming out as yesterday). Yet, the dates in the regular non-angular-ui bindings are coming out correct.

Is there an easy solution to this? Is there some way that I can force the input display dates in UTC?

**Comment by boneskull** [comment]
I have a fix for this.  Will push soon.

**Comment by boneskull** [comment]
See https://github.com/angular-ui/angular-ui/pull/163

**Comment by petebacondarwin** [comment]
Oh ok I get it. If you pass in a string rather than a real date object to
the directive you want it to parse the string using the specified date
format.
Christopher can you make the unit test more clear on this?

Pete
...from my mobile.

**Comment by boneskull** [comment]
I'm not sure what you want me to do, write a comment?

**Comment by petebacondarwin** [comment]
Hi Christopher
Sorry. I didn't mean to offend you.
The description and expectations should be enough without comments!
I just wasn't clear from the change and unit test what you were fixing. Is
it really to do with Iso dates?
Pete
...from my mobile.

**Comment by boneskull** [comment]
No offense taken, sorry if it sounded that way, I just didn't understand what you wanted.

No, the iso date thing was just a way to grab the string representation of the Date coming out of the datepicker() function which matched up pretty well with the string date that was passed in.  I guess this test is just asserting that the string date that is passed in vaguely matches the string date that comes out of the directive.  If we convert to Date within the directive, this test fails because the Date coming out of datepicker() will be different.

**Comment by bradcavanagh** [comment]
I've forked from master, then applied this patch, and it appears that either this patch doesn't fix the issue, as the new unit test fails:

```
Firefox 14.0 uiDate should parse the date correctly from a string FAILED
Expected 'Sun May 19 2019' to equal 'Mon Jul 30 2012'.
```

Without this patch the unit test still fails, which appears to duplicate the behaviour I reported:

```
Firefox 14.0 uiDate should retain timezone information FAILED
Expected '2012-08-16' to equal '2012-08-17'.
```

I'll try to see if I can figure out a fix, but I've never looked at coffeescript before, and date handling has always made me go wobbly.

**Comment by petebacondarwin** [comment]
Can you try testing this version in this punk: http://plnkr.co/edit/E2PJJ8
I don't have access to git or testacular to test it out.  I hope to replace
the coffee script date directive with this javascript one when I get back
from vacation.
It could well fail the "uiDate should parse the date correctly from a
string" test too but this test could well be flawed since how should the
directive know how to parse a date string?
It should be more like this:

it 'should parse the date correctly from a string', ()->
      inject ($compile, $rootScope)->
        aDateString = "2012-07-30T17:59:16.395Z"
        aDate = new Date(aDateString)
        element = $compile("<input ui-date='{ dateFormat=\"YY-m-d\"}'
ng-model='x'></input>")($rootScope)
        $rootScope.$apply ()->
          $rootScope.x = aDateString
        expect(element.datepicker('getDate').toDateString()).toEqual(aDate.toDateString())
        expect(element.datepicker('getDate').toTimeString()).not.toEqual(aDate.toTimeString())

This plunk version relies on the dateFormat setting to decide how to parse
a string.  It then always converts any string to a Date object.
Regards,
Pete

... sent from my tablet

**Comment by bradcavanagh** [comment]
I think that works for me. At least, it does at the current time (14:18 PDT, or 21:18 UTC). I don't know how it'll behave when I'm in today and UTC is tomorrow.

It does, however, fix another issue I was having in clearing out dates (previously reported in issue #167).

I've kind of worked around this by using Date objects, which seems to clear things up. I'm not sure if this is a really good workaround though, as I haven't tested it when UTC's date is tomorrow.

**Comment by boneskull** [comment]
That sounds good.

Are we in agreement that the time zone test is valid and it should pass?

I'm kind of confused about what's going on with this issue.

**Comment by boneskull** [comment]
Trying your plunker version there are some failing tests that I'm unsure how to resolve.  We can ignore the second one since it should be corrected by your new test.  But yeah… I don't know what's going on here.  Any idea?  When are you back from vacation?  :D

Chrome 19.0 uiDate simple use on input element should put the date in the model FAILED
    TypeError: Cannot read property 'selectedDay' of undefined
        at Datepicker._formatDate (/Volumes/projects/angular-ui/test/lib/jquery/jquery-ui-1.8.18.js:9002:26)
        at Datepicker._selectDate (/Volumes/projects/angular-ui/test/lib/jquery/jquery-ui-1.8.18.js:8193:47)
        at /Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:7:25
        at /Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:34:9
        at Object.invoke (/Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular.js:2795:28)
        at workFn (/Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular-mocks.js:1711:20)
        at /Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular-mocks.js:1698:30
        at [object Object].<anonymous> (/Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:30:7)
    Error: Declaration Location
        at /Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular-mocks.js:1697:17
        at [object Object].<anonymous> (/Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:30:7)

Chrome 19.0 uiDate should parse the date correctly from a string FAILED
    Error: Expected 'Mon May 20 2019' to equal 'Mon Jul 30 2012'.
        at /Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:48:60
        at Object.invoke (/Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular.js:2795:28)
        at workFn (/Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular-mocks.js:1711:20)
        at /Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular-mocks.js:1698:30
        at [object Object].<anonymous> (/Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:40:5)

Chrome 19.0 uiDate use with ng-required directive should be valid after the date has been picked FAILED
    TypeError: Cannot read property 'selectedDay' of undefined
        at Datepicker._formatDate (/Volumes/projects/angular-ui/test/lib/jquery/jquery-ui-1.8.18.js:9002:26)
        at Datepicker._selectDate (/Volumes/projects/angular-ui/test/lib/jquery/jquery-ui-1.8.18.js:8193:47)
        at /Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:7:25
        at /Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:88:9
        at Object.invoke (/Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular.js:2795:28)
        at workFn (/Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular-mocks.js:1711:20)
        at /Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular-mocks.js:1698:30
        at [object Object].<anonymous> (/Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:84:7)
    Error: Declaration Location
        at /Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular-mocks.js:1697:17
        at [object Object].<anonymous> (/Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:84:7)

Chrome 19.0 uiDate simple use on a div element should put the date in the model FAILED
    TypeError: Cannot read property 'selectedDay' of undefined
        at Datepicker._formatDate (/Volumes/projects/angular-ui/test/lib/jquery/jquery-ui-1.8.18.js:9002:26)
        at Datepicker._selectDate (/Volumes/projects/angular-ui/test/lib/jquery/jquery-ui-1.8.18.js:8193:47)
        at /Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:7:25
        at /Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:117:9
        at Object.invoke (/Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular.js:2795:28)
        at workFn (/Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular-mocks.js:1711:20)
        at /Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular-mocks.js:1698:30
        at [object Object].<anonymous> (/Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:113:7)
    Error: Declaration Location
        at /Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular-mocks.js:1697:17
        at [object Object].<anonymous> (/Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:113:7)

Chrome 19.0 uiDate use with ng-required directive should be valid after the date has been picked FAILED
    TypeError: Cannot read property 'selectedDay' of undefined
        at Datepicker._formatDate (/Volumes/projects/angular-ui/test/lib/jquery/jquery-ui-1.8.18.js:9002:26)
        at Datepicker._selectDate (/Volumes/projects/angular-ui/test/lib/jquery/jquery-ui-1.8.18.js:8193:47)
        at /Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:7:25
        at /Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:147:9
        at Object.invoke (/Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular.js:2795:28)
        at workFn (/Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular-mocks.js:1711:20)
        at /Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular-mocks.js:1698:30
        at [object Object].<anonymous> (/Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:143:7)
    Error: Declaration Location
        at /Volumes/projects/angular-ui/test/lib/angular-1.0.1/angular-mocks.js:1697:17
        at [object Object].<anonymous> (/Volumes/projects/angular-ui/modules/directives/date/test/dateSpec.js:143:7)

Chrome 19.0: Executed 5 of 130 (5 FAILED) (skipped 125) (0.369 / 0.042 secs) Use --force to continue. </WARN>

**Comment by petebacondarwin** [comment]
Hi Christopher
I believe the problem is that these tests don't allow the watches to run
that now initialize the query date pick. Previously the initialization was
done straight on the linking fn now it is done inside a watch handler. This
makes no difference in practise but blats the tests because they check
immediately for the widget. I think this should be solved by use of $apply
before the expectations are checked in the unit tests.

Pete
...from my mobile.

**Comment by boneskull** [comment]
I'll give that a try, thanks.

**Comment by petebacondarwin** [comment]
I am back on 1 Sept

Pete
...from my mobile.

**Comment by pkozlowski-opensource** [comment]
Fixed in PR #177.
